spec/blocksync/verification.md

```diff
@@ -0,0 +1,120 @@
+
+Currently there are two possible ways to obtain a trusted state :
+ 1. If the node is blocksync-ing from genesis, we have to verify the very first block and mark it as the initial trusted state.
+ 
+ [//]: # (trust the validators provided in the initial state and use them to verify the initial block received from peers.)
+
+ 2. The node loads the last block stored in its local block store. 
+
+**Case 1 - blocksync from genesis**
+
+In this case we only verify that the validator hash of the block matches the validator set of the initial state. We will use the initially provided validator set for verification and further verify the block against witne+sses. In this scenario, this state is very likely to be outside the trusting period. We will accept this block as a trusted state and store it inside the node's block store. 
+
+It is worth noting that, running block sync from the first height is significantly slower than running statesync first. However, statesync does not keep the entire blockchain history and some operators might opt not to state sync. The reason is that, if sufficiently many nodes state sync and other nodes who have historical data fail or leave the network, we have no history.
+
+**Case 2 - blocksync starting from pre-existing state**
+
+If the node is not starting from gensis, there will be pre-existing state in its local block store. 
+
+This can happen in the following two cases:
+
+- The node is blocksyncing after completing state sync.
+
+- The node has crashed and is now recovering.
+
+In both of these cases, we can trust the state in the store. State sync runs the light client verification protocol. If the crash happened during consensus or sync-ing, the only blocks in the store are blocks that are verified or decided. 
+
+If we can trust this initial state at height H, we can use this fact to eliminate faulty blocks whose `lastCommit` at height H+1 does not verify the trusted block. Additionally, we can double check the validator sets that signed the block at H + 1 against the expected validator sets for this height. The expected validator sets are stored within the trusted state and returned to us upon applying the block and executing it against the application.
+
+**Switching from statesync to blocksync**
+
+If we switch from statesync to blocksync, the last item in the block store is not the entire block. The store holds only the verified header and the commit that verified it. Therefore, we do need to fetch the corresponding block from our peers, verify it against data stored in the store and set the `last trusted block` to this block. 
+
+To enable this, upon switching from statesync, we do not instruct the block pool to fetch blocks at height `H + 1` (where height `H` corresponds to the height of the header last verified with statesync). Instead we let it fetch the block at height H, but do not execute this block against the application (as this would fail - state is already updated to height H + 1. )
```

Comment:
I don't understand why we need to fetch the whole block `H`. Isn't the App state already reflecting that height?
Shouldn't it be enough fetching `H+1` directly and verifying it against `H` (which is trusted via light client)?

Reply:
Right now, the verification of a block against another block, requires some fields not present in the header. We could alternatively not fetch block at height H in this particular scenario but verify only the signatures. But for H + 1 ,we do not have the signatures (they are stored at H + 2), so we would have to verify the signatures in `VerifyCommit` against H and decide to trust on H + 1 based on that. This is also a valid possibility.